File: src/api/vocabulary.ts
import { invoke } from "@tauri-apps/api/core";

// Retry configuration
const MAX_RETRIES = 3;
const INITIAL_DELAY = 100; // ms

// Helper function to sleep for a given duration
const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

// Helper function to check if error is a database lock error
const isDatabaseLockError = (error: any): boolean => {
  const errorStr = error?.toString() || '';
  const errorDetails = error?.details || error?.message || '';
  return errorStr.includes('database is locked') || 
         errorStr.includes('SQLITE_BUSY') ||
         errorDetails.includes('database is locked') ||
         errorDetails.includes('SQLITE_BUSY');
};

// Wrapper function to add retry logic with exponential backoff
async function invokeWithRetry<T>(
  command: string,
  args: Record<string, any>,
  retries: number = MAX_RETRIES
): Promise<T> {
  let lastError: any;
  
  for (let i = 0; i <= retries; i++) {
    try {
      return await invoke<T>(command, args);
    } catch (error) {
      lastError = error;
      
      // Only retry on database lock errors
      if (!isDatabaseLockError(error) || i === retries) {
        throw error;
      }
      
      // Exponential backoff: 100ms, 200ms, 400ms
      const delay = INITIAL_DELAY * Math.pow(2, i);
      console.log(`Database locked, retrying in ${delay}ms... (attempt ${i + 1}/${retries + 1})`);
      await sleep(delay);
    }
  }
  
  throw lastError;
}

export interface VocabularyItem {
  id?: string;
  user_id: string;
  video_id: string;
  word: string;
  timestamp: number;
  before_2_en?: string;
  before_2_zh?: string;
  before_2_timestamp?: number;
  before_1_en?: string;
  before_1_zh?: string;
  target_en: string;
  target_zh: string;
  dictionary_response?: string;
  review_stage?: number;
  next_review_at: string;
  last_reviewed_at?: string;
  is_phrase?: boolean;
  created_at?: string;
  scheduled_review_at?: string;
  review_count?: number;
  consecutive_correct?: number;
  was_late?: boolean;
  ever_overdue?: boolean;
  correct_count?: number;
  word_start_index?: number;
  word_end_index?: number;
}

export interface CreateVocabularyRequest {
  user_id: string;
  video_id: string;
  word: string;
  timestamp: number;
  before_2_en?: string;
  before_2_zh?: string;
  before_2_timestamp?: number;
  before_1_en?: string;
  before_1_zh?: string;
  target_en: string;
  target_zh: string;
  dictionary_response?: string;
  next_review_at: string;
  is_phrase?: boolean;
  word_start_index?: number;
}

export interface AccuracyStats {
  total_reviews: number;
  total_correct: number;
  accuracy_percentage: number;
  words_reviewed: number;
}

export const vocabularyApi = {
  async create(vocabulary: CreateVocabularyRequest): Promise<VocabularyItem> {
    return await invokeWithRetry<VocabularyItem>("create_vocabulary", { request: vocabulary });
  },

  async getByVideo(videoId: string, userId: string): Promise<VocabularyItem[]> {
    return await invokeWithRetry<VocabularyItem[]>("get_vocabulary_by_video", { videoId, userId });
  },

  async getAll(userId: string): Promise<VocabularyItem[]> {
    return await invokeWithRetry<VocabularyItem[]>("get_all_vocabulary", { userId });
  },

  async updateReview(vocabularyId: string, reviewStage: number, nextReviewAt: string): Promise<void> {
    return await invokeWithRetry<void>("update_vocabulary_review", { vocabularyId, reviewStage, nextReviewAt });
  },

  async delete(vocabularyId: string): Promise<void> {
    return await invokeWithRetry<void>("delete_vocabulary", { vocabularyId });
  },

  async getDueForReview(userId: string): Promise<VocabularyItem[]> {
    return await invokeWithRetry<VocabularyItem[]>("get_vocabulary_due_for_review", { userId });
  },

  async getDueForReviewByVideo(userId: string, videoId: string): Promise<VocabularyItem[]> {
    return await invokeWithRetry<VocabularyItem[]>("get_vocabulary_due_for_review_by_video", { userId, videoId });
  },

  async updateReviewWithResult(vocabularyId: string, isCorrect: boolean): Promise<VocabularyItem> {
    return await invokeWithRetry<VocabularyItem>("update_vocabulary_review_with_result", { vocabularyId, isCorrect });
  },

  async getOverdue(userId: string): Promise<VocabularyItem[]> {
    return await invokeWithRetry<VocabularyItem[]>("get_overdue_vocabulary", { userId });
  },

  async getAccuracyStats(userId: string): Promise<AccuracyStats> {
    return await invokeWithRetry<AccuracyStats>("get_vocabulary_accuracy_stats", { userId });
  }
};

File: src-tauri/src/models/vocabulary.rs
use diesel::prelude::*;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Queryable, Selectable, Serialize, Deserialize, Debug)]
#[diesel(table_name = crate::schema::vocabulary)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct Vocabulary {
    pub id: Option<String>,
    pub user_id: String,
    pub video_id: String,
    pub word: String,
    pub timestamp: i32,
    pub before_2_en: Option<String>,
    pub before_2_zh: Option<String>,
    pub before_1_en: Option<String>,
    pub before_1_zh: Option<String>,
    pub target_en: String,
    pub target_zh: String,
    pub dictionary_response: Option<String>,
    pub review_stage: Option<i32>,
    pub next_review_at: String,
    pub last_reviewed_at: Option<String>,
    pub is_phrase: Option<bool>,
    pub created_at: Option<String>,
    pub before_2_timestamp: Option<i32>,
    pub scheduled_review_at: Option<String>,
    pub review_count: Option<i32>,
    pub consecutive_correct: Option<i32>,
    pub was_late: Option<bool>,
}

#[derive(Insertable, Deserialize)]
#[diesel(table_name = crate::schema::vocabulary)]
pub struct NewVocabulary {
    pub id: String,
    pub user_id: String,
    pub video_id: String,
    pub word: String,
    pub timestamp: i32,
    pub before_2_en: Option<String>,
    pub before_2_zh: Option<String>,
    pub before_2_timestamp: Option<i32>,
    pub before_1_en: Option<String>,
    pub before_1_zh: Option<String>,
    pub target_en: String,
    pub target_zh: String,
    pub dictionary_response: Option<String>,
    pub review_stage: i32,
    pub next_review_at: String,
    pub last_reviewed_at: Option<String>,
    pub is_phrase: bool,
    pub scheduled_review_at: Option<String>,
    pub review_count: i32,
    pub consecutive_correct: i32,
    pub was_late: bool,
}

#[derive(Deserialize)]
pub struct CreateVocabularyRequest {
    pub user_id: String,
    pub video_id: String,
    pub word: String,
    pub timestamp: i32,
    pub before_2_en: Option<String>,
    pub before_2_zh: Option<String>,
    pub before_2_timestamp: Option<i32>,
    pub before_1_en: Option<String>,
    pub before_1_zh: Option<String>,
    pub target_en: String,
    pub target_zh: String,
    pub dictionary_response: Option<String>,
    pub next_review_at: String,
    pub is_phrase: Option<bool>,
}

impl CreateVocabularyRequest {
    pub fn into_new_vocabulary(self) -> NewVocabulary {
        let next_review_at = self.next_review_at.clone();
        NewVocabulary {
            id: Uuid::new_v4().to_string(),
            user_id: self.user_id,
            video_id: self.video_id,
            word: self.word,
            timestamp: self.timestamp,
            before_2_en: self.before_2_en,
            before_2_zh: self.before_2_zh,
            before_2_timestamp: self.before_2_timestamp,
            before_1_en: self.before_1_en,
            before_1_zh: self.before_1_zh,
            target_en: self.target_en,
            target_zh: self.target_zh,
            dictionary_response: self.dictionary_response,
            review_stage: 0,
            next_review_at: next_review_at.clone(),
            last_reviewed_at: None,
            is_phrase: self.is_phrase.unwrap_or(false),
            scheduled_review_at: Some(next_review_at),
            review_count: 0,
            consecutive_correct: 0,
            was_late: false,
        }
    }
}